packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,375 @@
+import { Tags } from "@waku/interfaces";
+import debug from "debug";
+import type { Libp2p } from "libp2p";
+
+import { KeepAliveManager, KeepAliveOptions } from "./keep_alive_manager.js";
+
+export const DEFAULT_MAX_BOOTSTRAP_PEERS_ALLOWED = 1;
+// TODO: add this functionality back
+// const DEFAULT_APPROX_TIME_BEFORE_BOOTSTRAP_FALLBACK_MS = 5000;
+export const DEFAULT_MAX_DIAL_ATTEMPTS_FOR_PEER = 3;
+export const DEFAULT_LOGGING_INTERVAL_MS = 10_000;
+
+const log = debug("waku:connection-manager");
+
+export interface Libp2pComponents {
+  connectionManager: Libp2p["connectionManager"];
+  peerStore: Libp2p["peerStore"];
+  ping: Libp2p["ping"];
+  dial: Libp2p["dial"];
+}
+
+export interface UpdatedStates {
+  allPeers: Peer[];
+  allDialablePeers: Peer[];
+  bootstrapPeers: Peer[];
+  nonBootstrapPeers: Peer[];
+  dialableBootstrapPeers: Peer[];
+  dialableNonBootstrapPeers: Peer[];
+  allConnections: Connection[];
+  bootstrapConnections: Connection[];
+  nonBootstrapConnections: Connection[];
+}
+
+export interface ConnectionManagerOptions {
+  /**
+   * Number of attempts before a peer is considered non-dialable
+   * This is used to not spam a peer with dial attempts when it is not dialable
+   */
+  maxDialAttemptsForPeer?: number;
+  /**
+   * Max number of bootstrap peers allowed to be connected to, initially
+   * This is used to increase intention of dialing non-bootstrap peers, found using other discovery mechanisms (like Peer Exchange)
+   */
+  maxBootstrapPeersAllowed?: number;
+  // TODO: add this functionality back
+  // /**
+  //  * Number of attempts before dialling bootstrap nodes over the {@link maxBootstrapPeersAllowed} value.
+  //  * Dialing is the process of opening an outgoing connection to a listening peer with a transport that is supported by both peers.
+  //  * This is only used when other discovery mechanisms haven't yield peers that are dialable
+  //  * This increases relative centralization of the network
+  //  */
+  // maxTimeBeforeBootstrapFallbackMs?: number;
+  /**
+   * Interval at which to log the state of the connection manager
+   */
+  loggingIntervalMs?: number;
+}
+
+export class ConnectionManager {
+  private static instances = new Map<string, ConnectionManager>();
+  private keepAliveManager: KeepAliveManager;
+  private options: ConnectionManagerOptions;
+  private libp2pComponents: Libp2pComponents;
+  private dialAttemptsForPeer: Map<string, number> = new Map();
+
+  public static create(
+    peerId: string,
+    libp2p: Libp2pComponents,
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay,
+    options?: ConnectionManagerOptions
+  ): ConnectionManager {
+    let instance = ConnectionManager.instances.get(peerId);
+    if (!instance) {
+      instance = new ConnectionManager(
+        libp2p,
+        keepAliveOptions,
+        relay,
+        options
+      );
+    }
+    ConnectionManager.instances.set(peerId, instance);
+
+    return instance;
```

Comment:
The only way for a library consumer to set their connection manager is as follows (please correct me if I'm mistaken):

```
const waku = new Waku(..) // this constructs a first connectionManager, let's call it A.
const B = ConnectionManager.create(...) // this is our new connection manager.
// Now, B has taken A's place in `ConnectionManager.instances`.
waku.connectionManager = B // now, there are no more references to A.
```

Right?

So in this case, all references to `A` have been removed. Does it mean `A` has been deleted and all the event listeners are stopped?

If so, please create a test to prove that this piece of code does what it's meant to do (ie, when 2 connection managers are created for the same libp2p instance).